torchinductor/scheduler.py

```diff
@@ -1000,188 +761,279 @@ def dead_node_elimination(self):
+            node.recursive_predecessors = recursive_predecessors
+
+        for order, node in enumerate(self.nodes):
+            node.min_order = order
+            node.max_order = order
+
+    def fuse_nodes(self):
+        """
+        Mutates self.nodes to combine nodes into FusedSchedulerNodes.
+        """
+        for _ in range(config.fusion_passes):
+            old_len = len(self.nodes)
+            self.fuse_nodes_once()
+            if len(self.nodes) == old_len:
+                break
+
+    def fuse_nodes_once(self):
+        """
+        Mutates self.nodes to combine nodes into FusedSchedulerNodes.
+
+        This relies on two key functions to control the logic:
+            - self.can_fuses(): checks if a fusion is legal
+            - self.score_fusion(): assigns priority to a given fusion
+        """
+        fused_nodes = set(self.nodes)
+        for node1, node2 in self.get_possible_fusions():
+            node1 = self.name_to_fused_node[node1.get_first_name()]
+            node2 = self.name_to_fused_node[node2.get_first_name()]
+            if self.can_fuse(node1, node2) and not self.will_fusion_create_cycle(
+                node1, node2
+            ):
+                node3 = FusedSchedulerNode.fuse(node1, node2)
+                fused_nodes.remove(node1)
+                fused_nodes.remove(node2)
+                fused_nodes.add(node3)
+                self.name_to_fused_node.update(
+                    {n.get_name(): node3 for n in node3.get_nodes()}
+                )
+        self.nodes = sorted(fused_nodes, key=lambda x: x.min_order)
+        self.topological_sort_schedule()
+
+    def get_possible_fusions(self):
+        """
+        Helper to find all legal fusion opportunities, sorted by self.score_fusion()
+        """
+        possible_fusions = []
+        for node1_index, node1 in enumerate(self.nodes):
+            for node2 in self.nodes[node1_index + 1 :]:
+                if self.can_fuse(node1, node2):
+                    possible_fusions.append((node1, node2))
+                elif self.can_fuse(node2, node1):
+                    # epilogue fusions are order dependent
+                    possible_fusions.append((node2, node1))
+        return sorted(possible_fusions, key=self.score_fusion_key, reverse=True)
+
+    def will_fusion_create_cycle(self, node1, node2):
+        """Finds whether there's a path from src to dst caused indirectly by fusion"""
+
+        def check(node):
+            if isinstance(node, FusedSchedulerNode) and node not in visited:
+                visited.add(node)
+                return bool(combined_names & node.recursive_predecessors) or any(
+                    check(self.name_to_fused_node[n])
+                    for n in node.recursive_predecessors - combined_predecessors
                 )
+            return False
+
+        visited = set()
+        combined_names = node1.get_names() | node2.get_names()
+        combined_predecessors = (
```

Comment:
Why can't you just do?
```
combined_names = node1.get_names() | node2.get_names()
combined_predecessors = node1.recursive_predcessors | node2.recursive_predecessors
return combined_names & combined_predecessors
```
?

Reply:
I think here we *dont* want to check combined_names, we want to exclude them with the set subtraction.  Nodes in the combined fusion can't create a cycle.